### PR TITLE
Import CSV parser library

### PR DESCRIPTION
**Pull request by vadmeste**
## Description
The CSV library code is imported from Golang 1.13.6

## Motivation and Context
Merge this first before we will add support of some S3 select configurations which are hardcoded in golang csv parser, such as, quote character, etc..

## How to test this PR?
No testing needed

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#52f6e71d-fe1f-41cb-b66b-85111e183daa) |
| mint-gateway-azure.sh | :x: [more...](#faaeb161-2ca1-4ca6-881b-fb2caebb289d) |
<a name="faaeb161-2ca1-4ca6-881b-fb2caebb289d"></a>
### 8927-ec0e5ef/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:31396
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 7caf969bf1a8:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 53 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 17 seconds
(4/13) Running mc tests ... done in 7 minutes and 12 seconds
(5/13) Running minio-go tests ... FAILED in 19 minutes and 46 seconds
{
  "args": {
    "bucketName": "minio-go-test-yu25uek2si2lqj6a",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 284,
  "error": "The specified bucket is not valid.",
  "function": "ListObjects(bucketName, objectPrefix, recursive, doneCh)",
  "message": "PutObject 1 call failed",
  "name": "minio-go: testListObjects",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
<a name="52f6e71d-fe1f-41cb-b66b-85111e183daa"></a>
### 8927-ec0e5ef/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:31604
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 49c918636c5b:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 25 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 7 seconds
(4/13) Running mc tests ... done in 52 seconds
(5/13) Running minio-go tests ... FAILED in 13 minutes and 39 seconds
{
  "args": {},
  "duration": 116926,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#7864211c-3d7b-4061-89fb-3cf2d70360c5) |
| mint-gateway-azure.sh | :x: [more...](#e1247e1d-9c3f-41b4-866c-eaa1b3146c67) |
<a name="7864211c-3d7b-4061-89fb-3cf2d70360c5"></a>
### 8927-1933b4b/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:31911
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 19e516753942:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 24 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 6 seconds
(4/13) Running mc tests ... done in 47 seconds
(5/13) Running minio-go tests ... FAILED in 14 minutes and 45 seconds
{
  "args": {},
  "duration": 119273,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
<a name="e1247e1d-9c3f-41b4-866c-eaa1b3146c67"></a>
### 8927-1933b4b/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:32301
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a085aabb5e1e:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 3 minutes and 20 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 18 seconds
(4/13) Running mc tests ... done in 6 minutes and 32 seconds
(5/13) Running minio-go tests ... FAILED in 32 minutes and 16 seconds
{
  "args": {
    "bucketName": "minio-go-test-xlwxvijah66ggdqv",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 234,
  "error": "The specified bucket is not valid.",
  "function": "ListObjects(bucketName, objectPrefix, recursive, doneCh)",
  "message": "PutObject 1 call failed",
  "name": "minio-go: testListObjects",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#337b894c-cf09-424a-89bc-dd420e451e27) |
| mint-worm.sh | :x: [more...](#5976f63c-a537-4d0e-81fe-5580b43c5efe) |
| mint-fs.sh | :x: [more...](#72062c4b-7f81-4451-b755-ee85e82e8151) |
| mint-gateway-azure.sh | :x: [more...](#05ad4b1d-450d-448b-b3f3-90b645183723) |
<a name="05ad4b1d-450d-448b-b3f3-90b645183723"></a>
### 8927-11ec4e0/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:31635
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 5f27266c27d9:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 57 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 16 seconds
(4/13) Running mc tests ... done in 6 minutes and 16 seconds
(5/13) Running minio-go tests ... FAILED in 20 minutes and 41 seconds
{
  "args": {
    "bucketName": "minio-go-test-2l3a1iwrzztp0vgz",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 238,
  "error": "The specified bucket is not valid.",
  "function": "ListObjects(bucketName, objectPrefix, recursive, doneCh)",
  "message": "PutObject 1 call failed",
  "name": "minio-go: testListObjects",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
<a name="72062c4b-7f81-4451-b755-ee85e82e8151"></a>
### 8927-11ec4e0/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:30113
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp f321f97e9b65:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 5 minutes and 10 seconds
{"name": "awscli", "duration": 307592, "function": "make_bucket\n", "status": "FAIL", "error": "Connect timeout on endpoint URL: "http://72.28.97.55:30113/awscli-mint-test-bucket-26450""}

Executed 0 out of 13 tests successfully.
```
<a name="337b894c-cf09-424a-89bc-dd420e451e27"></a>
### 8927-11ec4e0/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:32241
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 162b647f406b:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 21 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 7 seconds
(4/13) Running mc tests ... done in 48 seconds
(5/13) Running minio-go tests ... FAILED in 13 minutes and 52 seconds
{
  "args": {},
  "duration": 86678,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
<a name="5976f63c-a537-4d0e-81fe-5580b43c5efe"></a>
### 8927-11ec4e0/mint-worm.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:32651
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            worm
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d59acae8a4a1:/mint/log /tmp/mint-logs'

(1/1) Running worm tests ... FAILED in 2 minutes and 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "bucket-1pydlfa8i1jfg8cwqsq8ujqp3x1c8vfqvjmcqlhb6r5owcsx97xqwt4",
    "expiry": 60000000000,
    "objectName": "testObject"
  },
  "duration": 120314,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.55:32651/bucket-1pydlfa8i1jfg8cwqsq8ujqp3x1c8vfqvjmcqlhb6r5owcsx97xqwt4: dial tcp 72.28.97.55:32651: i/o timeout",
  "function": "PutAndDelete",
  "message": "WORM_MODE ON  - CreateBucket Failed",
  "name": "test worm mode",
  "status": "FAIL"
}

Executed 0 out of 1 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#0d0b0cc6-b0c1-4b74-a882-f8c62d603672) |
| mint-gateway-azure.sh | :x: [more...](#bf008ef4-6799-42a9-951e-670e10d73a4b) |
<a name="bf008ef4-6799-42a9-951e-670e10d73a4b"></a>
### 8927-97622b8/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:30606
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 00e59b106fa2:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 52 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 17 seconds
(4/13) Running mc tests ... done in 7 minutes and 23 seconds
(5/13) Running minio-go tests ... FAILED in 19 minutes and 32 seconds
{
  "args": {
    "bucketName": "minio-go-test-4rp9a2a3rzfdvv2x",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 235,
  "error": "The specified bucket is not valid.",
  "function": "ListObjects(bucketName, objectPrefix, recursive, doneCh)",
  "message": "PutObject 1 call failed",
  "name": "minio-go: testListObjects",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
<a name="0d0b0cc6-b0c1-4b74-a882-f8c62d603672"></a>
### 8927-97622b8/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:30253
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 6d18e389e99b:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 26 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 5 seconds
(4/13) Running mc tests ... done in 55 seconds
(5/13) Running minio-go tests ... FAILED in 13 minutes and 50 seconds
{
  "args": {},
  "duration": 103993,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by klauspost** [comment]
Add me for review, when it is ready ☺️

**Review by klauspost — DISMISSED** [review]
But this doesn't do anything yet.

Add the changes we want and let's look at it then.

**Comment by fwessels** [comment]
Do we need the writer(_test).go code at all?

**Comment by harshavardhana** [comment]
We do use writer to write response @fwessels

**Comment by vadmeste** [comment]
@klauspost I thought it is easier to merge this first, I am going to send further short PRs after this

**Comment by harshavardhana** [comment]
> @klauspost I thought it is easier to merge this first, I am going to send further short PRs after this

Oh, I see so the intention is take the fork and then add additions on top? @vadmeste - Its not a bad idea. Unless @klauspost @fwessels have any objections.

**Comment by vadmeste** [comment]
> Oh, I see so the intention is take the fork and then add additions on top? @vadmeste - Its not a bad idea. Unless @klauspost @fwessels have any objections.

yes I have another PR and writing some mint tests just for 'quote_character' config

File: pkg/csvparser/example_test.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
+// Copyright 2015 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
```
Comment (klauspost) [inline pkg/csvparser/example_test.go]: This references the wrong LICENSE file.

Maybe include the Go license in this folder?

Comment (harshavardhana) [reply pkg/csvparser/example_test.go]: Or just mention https://golang.org/LICENSE

Comment (vadmeste) [reply pkg/csvparser/example_test.go]: Done, I added https://golang.org/LICENCE

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#8d9e275c-241f-433d-9127-78aeab310a6d) |
| mint-large-bucket.sh | :x: [more...](#10bfbd4e-426b-4c42-befb-80c15d29efb0) |
<a name="8d9e275c-241f-433d-9127-78aeab310a6d"></a>
### 8927-b1e6ed4/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:31473
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 6f23ca7cead7:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 7 seconds
{
  "name": "awscli",
  "duration": 5284,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (XMinioServerNotInitialized) when calling the CreateBucket operation (reached max retries: 4): Server not initialized, please try again."
}

Executed 0 out of 13 tests successfully.
```
<a name="10bfbd4e-426b-4c42-befb-80c15d29efb0"></a>
### 8927-b1e6ed4/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:32192
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp bcd1ba974dc0:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 25 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 7 seconds
(4/13) Running mc tests ... done in 48 seconds
(5/13) Running minio-go tests ... FAILED in 14 minutes and 54 seconds
{
  "args": {},
  "duration": 97946,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#70b309fd-5758-4585-bce1-981aaeef35b7) |
| mint-large-bucket.sh | :x: [more...](#1258dce6-ac20-4dfe-8d9d-16562464cf17) |
<a name="70b309fd-5758-4585-bce1-981aaeef35b7"></a>
### 8927-dc4902d/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:31932
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 410d958ab34c:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 8 seconds
{
  "name": "awscli",
  "duration": 6033,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (XMinioServerNotInitialized) when calling the CreateBucket operation (reached max retries: 4): Server not initialized, please try again."
}

Executed 0 out of 13 tests successfully.
```
<a name="1258dce6-ac20-4dfe-8d9d-16562464cf17"></a>
### 8927-dc4902d/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.58:31794
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp cbc036bb1bcb:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 24 seconds
(2/13) Running aws-sdk-java tests ... done in 2 seconds
(3/13) Running aws-sdk-ruby tests ... done in 5 seconds
(4/13) Running mc tests ... done in 42 seconds
(5/13) Running minio-go tests ... FAILED in 12 minutes and 20 seconds
{
  "args": {},
  "duration": 111519,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by vadmeste** [comment]
ping @klauspost & @fwessels, PTAL

**Review by harshavardhana — APPROVED** [review]

**Review by nitisht — APPROVED** [review]

**Pull request merged**